tls/s2n_signature_algorithms.c

```diff
@@ -204,21 +205,42 @@ int s2n_choose_sig_scheme_from_peer_preference_list(struct s2n_connection *conn,
     return 0;
 }
 
-int s2n_send_supported_signature_algorithms(struct s2n_stuffer *out)
+int s2n_send_supported_sig_scheme_list(struct s2n_connection *conn, struct s2n_stuffer *out)
 {
-    /* The array of hashes and signature algorithms we support */
-    uint16_t num_signature_schemes = s2n_supported_sig_scheme_pref_list_len;
-    uint16_t signature_schemes_size = num_signature_schemes * TLS_SIGNATURE_SCHEME_LEN;
+    const struct s2n_signature_preferences *signature_preferences = conn->config->signature_preferences;
+    notnull_check(signature_preferences);
 
-    GUARD(s2n_stuffer_write_uint16(out, signature_schemes_size));
+    GUARD(s2n_stuffer_write_uint16(out, s2n_supported_sig_scheme_list_size(conn)));
 
-    for (int i =  0; i < num_signature_schemes; i++) {
-        GUARD(s2n_stuffer_write_uint16(out, s2n_supported_sig_scheme_pref_list[i]->iana_value));
+    for (int i =  0; i < signature_preferences->count; i++) {
+        const struct s2n_signature_scheme *const scheme = signature_preferences->signature_schemes[i];
+        if (0 == s2n_signature_scheme_valid_to_offer(conn, scheme)) {
+            GUARD(s2n_stuffer_write_uint16(out, scheme->iana_value));
+        }
     }
 
     return 0;
 }
 
+int s2n_supported_sig_scheme_list_size(struct s2n_connection *conn)
+{
+    return s2n_supported_sig_schemes_count(conn) * TLS_SIGNATURE_SCHEME_LEN;
+}
+
+int s2n_supported_sig_schemes_count(struct s2n_connection *conn)
+{
+    const struct s2n_signature_preferences *signature_preferences = conn->config->signature_preferences;
+    notnull_check(signature_preferences);
+
+    int count = 0;
+    for (int i =  0; i < signature_preferences->count; i++) {
+        if (0 == s2n_signature_scheme_valid_to_offer(conn, signature_preferences->signature_schemes[i])) {
+            count ++;
+        }
+    }
+    return count;
+}
```

Comment:
`s2n_signature_preferences ::count` is a `uint8_t` .  do we need to overflow check that when assigning it from an array, and should we be using the same type in this count function?

Reply:
This count var should be a uint8_t.

When setting up preferences, I'm not so sure. We don't check for overflow with cipher preferences. Since customers can only select sets (and not create our own) I think we're pretty safe there.

If we wanted to support anywhere near 255 signature scheme options, there would also be other changes we'd need to make, such as limiting how many we send. At the moment we only accept 32 at a time before throwing an error. If we send more than 32, all our self-talk tests will fail :)